controllers/apps/transformer_component_upgrade.go

```diff
@@ -0,0 +1,175 @@
+		return nil
+	}
+
+	var parent *model.ObjectVertex
+	legacyFound := false
+
+	// remove the RSM object if found
+	exists, err := legacyCRDExists(transCtx.Context, graphCli)
+	if err != nil {
+		return err
+	}
+	if exists {
+		rsm := &legacy.ReplicatedStateMachine{}
+		if err := graphCli.Get(transCtx.Context, client.ObjectKeyFromObject(comp), rsm); err == nil {
+			legacyFound = true
+			parent = graphCli.Do(dag, nil, rsm, model.ActionDeletePtr(), parent)
+		} else if !apierrors.IsNotFound(err) {
+			return err
+		}
+
+		// remove xxx-rsm-env configmap
+		env := &corev1.ConfigMap{}
+		key := types.NamespacedName{
+			Namespace: comp.Namespace,
+			Name:      fmt.Sprintf("%s-rsm-env", comp.Name),
+		}
+		if err := graphCli.Get(transCtx.Context, key, env); err == nil {
+			legacyFound = true
+			parent = graphCli.Do(dag, nil, env, model.ActionDeletePtr(), parent)
+		} else if !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	// remove the StatefulSet object if found
+	sts := &appsv1.StatefulSet{}
+	if err := graphCli.Get(transCtx.Context, client.ObjectKeyFromObject(comp), sts); err == nil {
+		legacyFound = true
+		parent = graphCli.Do(dag, nil, sts, model.ActionDeletePtr(), parent, model.WithPropagationPolicy(client.PropagationPolicy(metav1.DeletePropagationOrphan)))
+	} else if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	// update pod & pvc & svc labels
+	objectList := []client.ObjectList{&corev1.PersistentVolumeClaimList{}, &corev1.ServiceList{}, &corev1.PodList{}}
+	ml := client.MatchingLabels{
+		constant.AppManagedByLabelKey:   constant.AppName,
+		constant.AppInstanceLabelKey:    transCtx.Cluster.Name,
+		constant.KBAppComponentLabelKey: synthesizeComp.Name,
+	}
+	inNS := client.InNamespace(comp.Namespace)
+	var revision string
+	for _, list := range objectList {
+		if err := graphCli.List(transCtx.Context, list, ml, inNS); err == nil {
+			items := reflect.ValueOf(list).Elem().FieldByName("Items")
+			l := items.Len()
+			for i := 0; i < l; i++ {
+				object := items.Index(i).Addr().Interface().(client.Object)
+				if _, ok := object.GetLabels()[rsmcore.WorkloadsManagedByLabelKey]; ok {
+					continue
+				}
+				if _, ok := object.(*corev1.Service); ok && !strings.HasSuffix(object.GetName(), "-headless") {
+					continue
+				}
+				legacyFound = true
+				object.SetOwnerReferences([]metav1.OwnerReference{})
+				object.GetLabels()[rsmcore.WorkloadsManagedByLabelKey] = rsmcore.KindInstanceSet
+				object.GetLabels()[rsmcore.WorkloadsInstanceLabelKey] = comp.Name
+				if _, ok := object.(*corev1.Pod); ok {
+					if revision == "" {
+						revision, err = buildRevision(synthesizeComp, compDef)
+						if err != nil {
+							return err
+						}
+					}
+					object.GetLabels()[appsv1.ControllerRevisionHashLabelKey] = revision
+				}
+				parent = graphCli.Do(dag, nil, object, model.ActionUpdatePtr(), parent)
+			}
+		}
+	}
+
+	if legacyFound {
+		comp.Status.ObservedGeneration = 0
```

Comment:
Will this conflict with other operations of 'comp' in the DAG?  and better to add a comment why set comp.Status.ObservedGeneration = 0 here